File: Divide and conquer/쿼드트리.cpp
#include<cstdio>
#include<algorithm>

int arr[65][65];

int quad(int a, int b, int N)
{
	bool flag = true;

	for (int i = a; i <a+N && flag; i++)
	{
		for (int j = b; j < b + N; j++)
		{
			if (arr[a][b] != arr[i][j])
			{
				flag = false;
				break;
			}
		}
		}

	if (flag==true)
	{
		printf("%d", arr[a][b]);
	}
	else
	{
		printf("(");
		quad(a, b, N / 2);//왼쪽 위
		quad(a, b + N / 2, N/2);//오른쪽 위
		quad(a + N / 2, b,N/2);//왼쪽 아래
		quad(a + N / 2, b + N / 2, N/2);//오른쪽 아래
		printf(")");
	}
	return 0;
}

int main()
{
	int N;
	scanf("%d", &N);
	for (int i = 1; i <= N; i++)
		for (int j = 1; j <= N; j++)
			scanf("%1d", &arr[i][j]);

	quad(1, 1, N);
	printf("\n");

	return 0;
}

File: Divide and conquer/쿼드트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {//1992

	private static int N;
	private static int[][] arr;

	public static void main(String[] args) throws NumberFormatException, IOException {


		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		N = Integer.parseInt(br.readLine());
		arr = new int[N][N];

		for(int i=0;i<N;i++){
			String readline = br.readLine();
			for(int j=0;j<N;j++){
				arr[i][j] = readline.charAt(j)-48;
			}
		}
		condiv(0,0,N);
		br.close();
	}
	public static void condiv(int start_x,int start_y,int size) {

		int first = arr[start_x][start_y];
		boolean flag =false;

		esc:  for(int i=start_x;i<start_x+size;i++){
			for(int j=start_y;j<start_y+size;j++){
				if(first!=arr[i][j]){
					flag = true;
					break esc;
				}
			}
		}
		if(flag){
			size/=2;
			System.out.print("(");
			condiv(start_x,start_y,size);//왼쪽 위 00~33
			condiv(start_x,start_y+size,size);//오른쪽 위 04~37
			condiv(start_x+size,start_y,size);//왼쪽 아래 40~73
			condiv(start_x+size,start_y+size,size);//오른쪽 아래 44~77
			System.out.print(")");
		}
		else{
			System.out.print(arr[start_x][start_y]);
		}
	}
}
